File: frontend/src/api/finance.ts
import api from "./http";
import type { PageResult } from "./logs";

export type FinancialLogRow = {
  id: number;
  customer_id?: number;
  owner_username?: string;
  node_id?: number;
  node_name?: string;
  remote_email?: string;
  customer_name: string;
  renew_price: string;
  amount?: number;
  renew_days: number;
  new_expiry: string;
  created_at: string;
};

export type FinancialLogUpdatePayload = {
  customer_name?: string;
  owner_username?: string;
  renew_price?: string;
  amount?: number;
  renew_days?: number;
  new_expiry?: string;
  created_at?: string;
};

export type FinancialLogPage = PageResult<FinancialLogRow> & {
  total_amount: number;
};

export async function fetchFinancialLogs(params: {
  page?: number;
  per_page?: number;
  keyword?: string;
  owner_username?: string;
  node_id?: number | string;
  date_from?: string;
  date_to?: string;
} = {}) {
  const cleanParams = Object.fromEntries(
    Object.entries(params).filter(([, value]) => value !== "" && value !== undefined && value !== null),
  );
  const response = await api.get("/api/v1/finance/logs", { params: cleanParams });
  return response.data.data as FinancialLogPage;
}

export async function updateFinancialLog(logId: number, payload: FinancialLogUpdatePayload) {
  const response = await api.put(`/api/v1/finance/logs/${logId}`, payload);
  return response.data as { success: boolean; message: string; data: FinancialLogRow };
}

export async function deleteFinancialLog(logId: number) {
  const response = await api.delete(`/api/v1/finance/logs/${logId}`);
  return response.data as { success: boolean; message: string };
}

File: backend/app/api/v1/finance.py
from typing import Any

from fastapi import APIRouter, Depends, Query, Request
from pydantic import BaseModel

from backend.app.core.config import Settings
from backend.app.core.deps import require_admin
from backend.app.services.finance import delete_financial_log, list_financial_logs, update_financial_log


router = APIRouter(prefix="/api/v1/finance", tags=["finance"])


class FinancialLogUpdatePayload(BaseModel):
    customer_name: str | None = None
    owner_username: str | None = None
    renew_price: str | None = None
    amount: float | None = None
    renew_days: int | None = None
    new_expiry: str | None = None
    created_at: str | None = None


@router.get("/logs")
def financial_logs(
    request: Request,
    page: int = 1,
    per_page: int = 20,
    keyword: str = "",
    owner_username: str = "",
    node_id: str | None = Query(default=None),
    date_from: str = "",
    date_to: str = "",
    username: str = Depends(require_admin),
):
    settings: Settings = request.app.state.settings
    clean_node_id = int(node_id) if str(node_id or "").strip() else None
    return {
        "success": True,
        "data": list_financial_logs(
            settings,
            page=page,
            per_page=per_page,
            keyword=keyword,
            owner_username=owner_username,
            node_id=clean_node_id,
            date_from=date_from,
            date_to=date_to,
        ),
    }


@router.put("/logs/{log_id}")
def update_log(request: Request, log_id: int, payload: FinancialLogUpdatePayload, username: str = Depends(require_admin)):
    settings: Settings = request.app.state.settings
    data: dict[str, Any] = {key: value for key, value in payload.model_dump().items() if value is not None}
    return update_financial_log(settings, log_id, data, actor=username)


@router.delete("/logs/{log_id}")
def delete_log(request: Request, log_id: int, username: str = Depends(require_admin)):
    settings: Settings = request.app.state.settings
    return delete_financial_log(settings, log_id, actor=username)
